File: server/src/engine/systems/worldsystem.ts
import EntityManager = require("../entitymanager");
import Tile = require("../world/tile");
import Vector2 = require("../geometry/vector2");
import WallFactory = require("../factories/wallfactory");
import Transform = require("../components/transform");

class WorldSystem{

    settings;
    tiles: Array<Array<Tile>>;
    private _positions: { [entity: number]: Vector2 } //entity -> position

    constructor(private _entityManager: EntityManager) {
        this.tiles = [];
        this._positions = {};
        this.settings = {
            tilesX: 20, //The number of horizontal tiles on this map
            tilesY: 10, //The number of vertical tiles on this map
        };
        this.initialize();
    }

    initialize() {
        for (var x = 0; x < this.settings.tilesX; x++) {
            this.tiles[x] = [];
            for (var y = 0; y < this.settings.tilesY; y++) {
                this.tiles[x][y] = new Tile();
                if ((y == 0) || (y == this.settings.tilesY - 1) || (x == 0) || (x == this.settings.tilesX - 1)) {
                    var wall = WallFactory.new(this._entityManager, new Vector2(x, y));
                    this.addEntity(new Vector2(x,y), wall);
                }
            }
        }
    }

    addEntity(position: Vector2, entity: number) {
        this._positions[entity] = position;
        var tile = this.tiles[position.x][position.y];
        var volume = this._entityManager.getComponent(entity, Transform).volume;
        tile.entities.push(entity);
        tile.volume += volume;
    }

    removeEntity(entity: number) {
        var position = this._positions[entity];
        var tile = this.tiles[position.x][position.y];
        var index = tile.entities.indexOf(entity);
        if (index === -1) {
            return;
        }
        tile.entities.splice(index, 1);
        var volume = this._entityManager.getComponent(entity, Transform).volume;
        tile.volume -= volume;
    }

    moveEntity(entity: number, position: Vector2) {
        this.removeEntity(entity);
        this.addEntity(position, entity);
    }
}

export = WorldSystem;

File: sharpServer/Engine/Systems/WorldSystem.cs
using System;
using System.Collections.Generic;

namespace SharpServer.Engine.Systems
{
    internal class WorldSystem
    {
        public int tilesX;
        public int tilesY;
        public Tile[,] tiles;
        private Dictionary<int, Vector2> positions; //entity -> position
        private EntityManager entityManager;

        public WorldSystem(EntityManager em)
        {
            positions = new Dictionary<int, Vector2>();
            entityManager = em;
            tilesX = 20;
            tilesY = 10;
            tiles = new Tile[tilesX, tilesY];
            initialize();
        }

        private void initialize()
        {
            for (var x = 0; x < tilesX; x++)
            {
                for (var y = 0; y < tilesY; y++)
                {
                    tiles[x, y] = new Tile();
                    if ((y == 0) || (y == tilesY - 1) || (x == 0) || (x == tilesX - 1))
                    {
                        var wall = Factories.WallFactory.newWall(entityManager, new Vector2(x, y));
                        addEntity(new Vector2(x, y), wall);
                    }
                }
            }
        }

        public void addEntity(Vector2 position, int entity)
        {
            positions[entity] = position;
            var tile = tiles[position.x, position.y];
            var volume = entityManager.getComponent<Shape>(entity).volume;
            tile.entities.Add(entity);
            tile.volume += volume;
        }

        public void removeEntity(int entity)
        {
            var position = positions[entity];
            var tile = this.tiles[position.x, position.y];
            if (!tile.entities.Contains(entity))
                throw new MissingMemberException("REMOVE ERROR: there is no entity " + entityManager.nameFor(entity) + "(ID: " + entity + ") in tile " + position.x + ":" + position.y);
            tile.entities.Remove(entity);
            var volume = entityManager.getComponent<Shape>(entity).volume;
            tile.volume -= volume;
        }

        public void moveEntity(int entity, Vector2 position)
        {
            removeEntity(entity);
            addEntity(position, entity);
        }
    }
}